File: python/services/identity/application/service.py
"""IdentityService — all identity use cases (A.3, A.8).

Depends only on domain types + ports. No infrastructure imports allowed
(enforced by import-linter, A.9).
"""
from __future__ import annotations

import secrets

from milton_prism.types.query_params.v1.query_params_pb2 import PageQueryParams

from services.identity.domain.domain import (
    AuthorizationTokens,
    Pagination,
    User,
    UsersFilter,
    UserState,
)
from services.identity.domain.errors import (
    ERR_ACCOUNT_SUSPENDED,
    ERR_INTERNAL,
    ERR_INVALID_CREDENTIALS,
    ERR_INVALID_SESSION,
    ERR_INVALID_TOKEN,
    ERR_MISSING_EMAIL,
    ERR_MISSING_IDENTIFIER,
    ERR_MISSING_PASSWORD,
    ERR_MISSING_PAYLOAD,
    ERR_TOKEN_GENERATION,
    ERR_TOKEN_REFRESH,
    ERR_USER_NOT_ACTIVE,
    ERR_USER_NOT_FOUND,
)
from services.identity.ports.auth import PasswordHasher, SessionStore, TokenManager
from services.identity.ports.repository import UserRepository
from services.identity.ports.transaction import TransactionManager
from shared.errors import DomainError
from shared.logging import infof


class IdentityService:
    """Orchestrates identity use cases. Assembled in wire.py (A.3)."""

    def __init__(
        self,
        repo: UserRepository,
        tx: TransactionManager | None,
        hasher: PasswordHasher,
        tokens: TokenManager,
        sessions: SessionStore,
    ) -> None:
        self._repo = repo
        self._tx = tx
        self._hasher = hasher
        self._tokens = tokens
        self._sessions = sessions

    async def create_user(
        self,
        user: User,
        plain_password: str,
    ) -> User:
        if not user.email:
            raise ERR_MISSING_PAYLOAD
        if not plain_password:
            raise ERR_MISSING_PASSWORD
        password_hash = self._hasher.hash(plain_password)
        try:
            created = await self._repo.create(user, password_hash)
        except DomainError:
            raise
        except Exception as exc:
            raise DomainError(ERR_INTERNAL.code, ERR_INTERNAL.message) from exc
        infof("identity.create-user", identifier=created.identifier)
        return created

    async def get_user(self, identifier: int) -> User:
        if identifier == 0:
            raise ERR_MISSING_IDENTIFIER
        try:
            return await self._repo.get_by_id(identifier)
        except DomainError:
            raise
        except Exception as exc:
            raise DomainError(ERR_INTERNAL.code, ERR_INTERNAL.message) from exc

    async def list_users(
        self,
        filter: UsersFilter | None,
        params: PageQueryParams | None = None,
    ) -> tuple[list[User], Pagination]:
        if params is None:
            params = PageQueryParams(page_number=1, page_size=10)
        try:
            return await self._repo.list_users(filter, params)
        except DomainError:
            raise
        except Exception as exc:
            raise DomainError(ERR_INTERNAL.code, ERR_INTERNAL.message) from exc

    async def update_user(
        self,
        user: User,
        update_mask: list[str],
    ) -> User:
        if user.identifier == 0:
            raise ERR_MISSING_IDENTIFIER
        try:
            existing = await self._repo.get_by_id(user.identifier)
        except DomainError:
            raise
        except Exception as exc:
            raise DomainError(ERR_INTERNAL.code, ERR_INTERNAL.message) from exc

        paths = update_mask if update_mask else ["email", "display_name", "system_user", "state"]
        for path in paths:
            if path == "email":
                existing.email = user.email
            elif path == "display_name":
                existing.display_name = user.display_name
            elif path == "system_user":
                existing.system_user = user.system_user
            elif path == "state":
                existing.state = user.state

        try:
            updated = await self._repo.update(existing)
        except DomainError:
            raise
        except Exception as exc:
            raise DomainError(ERR_INTERNAL.code, ERR_INTERNAL.message) from exc
        return updated

    async def delete_user(self, identifier: int) -> None:
        if identifier == 0:
            raise ERR_MISSING_IDENTIFIER
        try:
            await self._repo.soft_delete(identifier)
        except DomainError:
            raise
        except Exception as exc:
            raise DomainError(ERR_INTERNAL.code, ERR_INTERNAL.message) from exc

    async def authenticate_user(
        self,
        email: str,
        plain_password: str,
    ) -> AuthorizationTokens:
        if not email:
            raise ERR_MISSING_EMAIL
        if not plain_password:
            raise ERR_MISSING_PASSWORD

        try:
            user, password_hash = await self._repo.get_by_email(email)
        except DomainError as exc:
            # Any lookup error maps to generic invalid-credentials (hide user-existence)
            raise DomainError(
                ERR_INVALID_CREDENTIALS.code, ERR_INVALID_CREDENTIALS.message
            ) from exc
        except Exception as exc:
            raise DomainError(ERR_INTERNAL.code, ERR_INTERNAL.message) from exc

        if user.state == UserState.USER_STATE_DELETED:
            raise ERR_USER_NOT_ACTIVE
        if user.state == UserState.USER_STATE_SUSPENDED:
            raise ERR_ACCOUNT_SUSPENDED

        if not self._hasher.verify(password_hash, plain_password):
            raise ERR_INVALID_CREDENTIALS

        session_id = _generate_session_id()
        try:
            tokens = self._tokens.new_tokens(user.identifier, user.system_user, session_id)
        except Exception as exc:
            raise DomainError(ERR_TOKEN_GENERATION.code, ERR_TOKEN_GENERATION.message) from exc

        try:
            await self._sessions.save(session_id, user.identifier, user.system_user)
        except Exception as exc:
            raise DomainError(ERR_INTERNAL.code, ERR_INTERNAL.message) from exc

        infof("identity.authenticate-user", identifier=user.identifier)
        return tokens

    async def refresh_token(self, refresh_token_value: str) -> AuthorizationTokens:
        if not refresh_token_value:
            raise ERR_INVALID_TOKEN

        try:
            session_id = self._tokens.extract_session_id(refresh_token_value)
        except Exception as exc:
            raise DomainError(ERR_INVALID_TOKEN.code, ERR_INVALID_TOKEN.message) from exc

        try:
            user_id, _is_system, valid = await self._sessions.get(session_id)
        except Exception as exc:
            raise DomainError(ERR_INVALID_SESSION.code, ERR_INVALID_SESSION.message) from exc

        if not valid:
            raise DomainError(ERR_INVALID_SESSION.code, ERR_INVALID_SESSION.message)

        try:
            user = await self._repo.get_by_id(user_id)
        except DomainError as exc:
            raise DomainError(ERR_USER_NOT_FOUND.code, ERR_USER_NOT_FOUND.message) from exc
        except Exception as exc:
            raise DomainError(ERR_INTERNAL.code, ERR_INTERNAL.message) from exc

        try:
            new_tokens = self._tokens.new_tokens(
                user.identifier, user.system_user, session_id
            )
        except Exception as exc:
            raise DomainError(ERR_TOKEN_REFRESH.code, ERR_TOKEN_REFRESH.message) from exc

        return new_tokens

    async def logout(self, session_id: str) -> None:
        if not session_id:
            return
        try:
            await self._sessions.delete(session_id)
        except Exception as exc:
            raise DomainError(ERR_INTERNAL.code, ERR_INTERNAL.message) from exc
        infof("identity.logout", session_id=session_id)

    async def get_current_user(self, session_id: str) -> User:
        if not session_id:
            raise ERR_INVALID_TOKEN

        try:
            user_id, _is_system, valid = await self._sessions.get(session_id)
        except Exception as exc:
            raise DomainError(ERR_INVALID_SESSION.code, ERR_INVALID_SESSION.message) from exc

        if not valid:
            raise DomainError(ERR_INVALID_SESSION.code, ERR_INVALID_SESSION.message)

        try:
            return await self._repo.get_by_id(user_id)
        except DomainError:
            raise
        except Exception as exc:
            raise DomainError(ERR_INTERNAL.code, ERR_INTERNAL.message) from exc


    def extract_session_id_from_access_token(self, access_token: str) -> str:
        """Extract session_id from a bearer access token. Returns '' on failure.

        Called by handlers before delegating to logout/get_current_user so that
        handlers never need to access the token adapter directly.
        """
        if not access_token:
            return ""
        try:
            _, session_id = self._tokens.verify_access(access_token)
            return session_id
        except Exception:
            return ""


def _generate_session_id() -> str:
    """Generate a 16-byte cryptographically random hex session ID."""
    return secrets.token_hex(16)

File: core/services/identity/application/service.go
// Package application implements the use cases of the identity service.
//
// Service depends only on driven ports (repository, hasher, token manager,
// session store). All authentication side effects are behind those ports so
// the core is testable and free of MongoDB / Redis / gRPC client knowledge.
package application

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"errors"
	"fmt"

	"milton_prism/core/services/identity/domain"
	"milton_prism/core/services/identity/ports"
	applog "milton_prism/pkg/log"
	paginationv1 "milton_prism/pkg/pb/gen/milton_prism/types/pagination/v1"
	queryparamsv1 "milton_prism/pkg/pb/gen/milton_prism/types/query_params/v1"
	"milton_prism/pkg/pb/impl"

	"google.golang.org/protobuf/types/known/fieldmaskpb"
)

// Service orchestrates identity use cases.
type Service struct {
	repo     ports.UserRepository
	tx       ports.TransactionManager
	hasher   ports.PasswordHasher
	tokens   ports.TokenManager
	sessions ports.SessionStore
}

// NewService wires the identity application service.
func NewService(
	repo ports.UserRepository,
	tx ports.TransactionManager,
	hasher ports.PasswordHasher,
	tokens ports.TokenManager,
	sessions ports.SessionStore,
) *Service {
	return &Service{repo: repo, tx: tx, hasher: hasher, tokens: tokens, sessions: sessions}
}

// CreateUser persists a new user account with a hashed password.
func (s *Service) CreateUser(ctx context.Context, u *domain.User, plainPassword string) (*domain.User, error) {
	if u == nil || u.GetEmail() == "" {
		return nil, domain.ErrMissingPayload
	}
	if plainPassword == "" {
		return nil, domain.ErrMissingPassword
	}
	hash := plainPassword
	if s.hasher != nil {
		h, err := s.hasher.Hash(plainPassword)
		if err != nil {
			return nil, fmt.Errorf("%w: %v", domain.ErrInternal, err)
		}
		hash = h
	}
	out, err := s.repo.Create(ctx, u, hash)
	if err != nil {
		if errors.Is(err, domain.ErrEmailAlreadyExists) {
			return nil, err
		}
		return nil, fmt.Errorf("%w: %v", domain.ErrInternal, err)
	}
	return out, nil
}

// GetUser returns a user by identifier.
func (s *Service) GetUser(ctx context.Context, identifier uint64) (*domain.User, error) {
	if identifier == 0 {
		return nil, domain.ErrMissingIdentifier
	}
	out, err := s.repo.GetByID(ctx, identifier, false)
	if err != nil {
		if errors.Is(err, domain.ErrUserNotFound) {
			return nil, err
		}
		return nil, fmt.Errorf("%w: %v", domain.ErrInternal, err)
	}
	return out, nil
}

// ListUsers returns a paginated list of users.
func (s *Service) ListUsers(ctx context.Context, filter *domain.UsersFilter, params *queryparamsv1.PageQueryParams) ([]*domain.User, *paginationv1.Pagination, error) {
	if params == nil {
		params = impl.PageQueryParamsDefault()
	}
	impl.NormalizePaginationParams(params)
	items, p, err := s.repo.List(ctx, filter, params)
	if err != nil {
		return nil, nil, fmt.Errorf("%w: %v", domain.ErrInternal, err)
	}
	return items, p, nil
}

// UpdateUser applies a FieldMask-bounded partial update to a user.
// A nil or empty mask means "update all mutable fields".
func (s *Service) UpdateUser(ctx context.Context, u *domain.User, mask *fieldmaskpb.FieldMask) (*domain.User, error) {
	if u == nil || u.GetIdentifier() == 0 {
		return nil, domain.ErrMissingIdentifier
	}
	existing, err := s.repo.GetByID(ctx, u.GetIdentifier(), false)
	if err != nil {
		if errors.Is(err, domain.ErrUserNotFound) {
			return nil, err
		}
		return nil, fmt.Errorf("%w: %v", domain.ErrInternal, err)
	}
	paths := mask.GetPaths()
	if len(paths) == 0 {
		paths = []string{"email", "display_name", "system_user", "state"}
	}
	for _, p := range paths {
		switch p {
		case "email":
			existing.Email = u.GetEmail()
		case "display_name":
			existing.DisplayName = u.GetDisplayName()
		case "system_user":
			existing.SystemUser = u.GetSystemUser()
		case "state":
			existing.State = u.GetState()
		}
	}
	if err := s.repo.Update(ctx, existing); err != nil {
		if errors.Is(err, domain.ErrUserNotFound) {
			return nil, err
		}
		return nil, fmt.Errorf("%w: %v", domain.ErrInternal, err)
	}
	return existing, nil
}

// DeleteUser soft-deletes a user by setting delete_time.
func (s *Service) DeleteUser(ctx context.Context, identifier uint64) error {
	if identifier == 0 {
		return domain.ErrMissingIdentifier
	}
	if err := s.repo.SoftDelete(ctx, identifier); err != nil {
		if errors.Is(err, domain.ErrUserNotFound) {
			return err
		}
		return fmt.Errorf("%w: %v", domain.ErrInternal, err)
	}
	return nil
}

// AuthenticateUser validates credentials and returns fresh auth tokens.
// Password is never returned or persisted in plaintext.
func (s *Service) AuthenticateUser(ctx context.Context, email, plainPassword string) (*domain.AuthorizationTokens, error) {
	if email == "" {
		return nil, domain.ErrMissingEmail
	}
	if plainPassword == "" {
		return nil, domain.ErrMissingPassword
	}
	user, hash, err := s.repo.GetCredentialsByEmail(ctx, email)
	if err != nil {
		if errors.Is(err, domain.ErrUserNotFound) {
			return nil, domain.ErrInvalidCredentials
		}
		return nil, fmt.Errorf("%w: %v", domain.ErrInternal, err)
	}
	if user.GetState() == domain.UserStateDeleted {
		return nil, domain.ErrUserNotActive
	}
	if user.GetState() == domain.UserStateSuspended {
		return nil, domain.ErrAccountSuspended
	}
	if s.hasher != nil {
		if err := s.hasher.Verify(hash, plainPassword); err != nil {
			return nil, domain.ErrInvalidCredentials
		}
	}
	if s.tokens == nil {
		return nil, domain.ErrTokenGeneration
	}
	sessionID, err := generateSessionID()
	if err != nil {
		return nil, fmt.Errorf("%w: %v", domain.ErrInternal, err)
	}
	tokens, err := s.tokens.NewTokens(ctx, user.GetIdentifier(), user.GetSystemUser(), sessionID)
	if err != nil {
		return nil, domain.ErrTokenGeneration
	}
	if s.sessions != nil {
		if err := s.sessions.Save(ctx, sessionID, user.GetIdentifier(), user.GetSystemUser()); err != nil {
			return nil, fmt.Errorf("%w: %v", domain.ErrInternal, err)
		}
	}
	return tokens, nil
}

// RefreshToken validates the supplied refresh token and returns a new token pair.
func (s *Service) RefreshToken(ctx context.Context, rawRefreshToken string) (*domain.AuthorizationTokens, error) {
	if rawRefreshToken == "" {
		return nil, domain.ErrInvalidToken
	}
	if s.tokens == nil {
		return nil, domain.ErrTokenRefresh
	}
	sessionID, err := s.tokens.ExtractSessionID(rawRefreshToken)
	if err != nil {
		return nil, domain.ErrInvalidToken
	}
	userID, _, valid, err := s.sessions.Get(ctx, sessionID)
	if err != nil || !valid {
		return nil, domain.ErrInvalidSession
	}
	user, err := s.repo.GetByID(ctx, userID, false)
	if err != nil {
		return nil, domain.ErrUserNotFound
	}
	if err := s.tokens.Revoke(ctx, rawRefreshToken); err != nil {
		applog.Warningf("identity.RefreshToken: revoke old token: code=%s error=%v", domain.ErrCodeInternal, err)
	}
	tokens, err := s.tokens.NewTokens(ctx, user.GetIdentifier(), user.GetSystemUser(), sessionID)
	if err != nil {
		return nil, domain.ErrTokenRefresh
	}
	return tokens, nil
}

// Logout invalidates the user session and revokes the access token.
func (s *Service) Logout(ctx context.Context, sessionID, accessToken string) error {
	if s.sessions != nil && sessionID != "" {
		if err := s.sessions.Delete(ctx, sessionID); err != nil {
			return fmt.Errorf("%w: %v", domain.ErrInternal, err)
		}
	}
	if s.tokens != nil && accessToken != "" {
		if err := s.tokens.Revoke(ctx, accessToken); err != nil {
			return fmt.Errorf("%w: %v", domain.ErrInternal, err)
		}
	}
	return nil
}

// generateSessionID returns a 16-byte cryptographically random hex string used as session ID.
func generateSessionID() (string, error) {
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return hex.EncodeToString(b), nil
}
